README.adoc

```diff
@@ -56,148 +56,97 @@ The driver will use default credentials set in the Google Cloud SDK `gcloud` app
-
-NOTE: The lack of foreign key constraints also means database-side cascading deletes are not supported via the `@OnDelete(action = OnDeleteAction.CASCADE)` annotation because there is no `ON DELETE CASCADE` constraint in Cloud Spanner DDL.
-However, Hibernate-side cascading operations such as `@ManyToOne(cascade = {CascadeType.ALL})` are supported.
-
-==== Optimizing Queries
-
-To take advantage of Cloud Spanner's full performance, consider using its https://cloud.google.com/spanner/docs/sql-best-practices[best practices].
-
-Because Hibernate's built-in HQL does not account for Spanner-specific features, you can construct queries directly from Cloud Spanner SQL instead:
-
-[source,java]
-----
-SQLQuery query = session.createSQLQuery("SELECT * FROM Singers AS s
-                                         JOIN@{FORCE_JOIN_ORDER=TRUE} Albums AS a
-                                         ON s.SingerId = a.Singerid
-                                         WHERE s.LastName LIKE '%x%'
-                                         AND a.AlbumTitle LIKE '%love%';");
-----
-
-==== Interleaved Tables
-
-Currently there is no support for Cloud Spanner's interleaved table feature.
-Hibernate ORM relationships are always constructed using join-tables and foreign keys by the framework.
-If your application would benefit from interleaved tables, we recommend that you manually create them and use types mapped to them in Hibernate. This enables the performance and database-enforced-constraint benefits from interleaved tables in Cloud Spanner, even though the Hibernate framework will not be aware of their relationship or any restrictions on operations between interleaved tables.
-
-==== Constraints
-
-Cloud Spanner does not support database constraints.
-As a result, `SpannerDialect` does not currently support any constraints such as `FOREIGN KEY`, `UNIQUE`, or `ON DELETE CASCADE`.
-
-NOTE: Hibernate doesn't directly rely on the existence of constraints to perform its operations and leaves the enforcement of relationship links to the database.
-
-The lack of the foreign key constraint affects relationships and collection properties annotated with `@ElementCollection`.
-
-
-==== Generated IDs
-
-Hibernate's `@GeneratedValue` annotation for `long` properties is supported but not recommended:
 [source, java]
 ----
 @Entity
 public class Employee {
 
   @Id
-  @GeneratedValue // DON'T DO THIS, IT'S SLOW
-  public Long id;
+  @GeneratedValue(strategy = GenerationType.AUTO)
+  @Type(type="uuid-char")
+  public UUID id;
 }
 ----
 
-This results in sequential IDs that are https://cloud.google.com/spanner/docs/schema-and-data-model#primary_keys[not optimal for Cloud Spanner] and requires locking of the `hibernate_sequence` table for inserts.
+The `@Type(type="uuid-char")` annotation specifies that this UUID value will be stored in Cloud Spanner as a `STRING` column.
+Leaving out this annotation causes a `BYTES` column to be used.
```

Comment:
Yeah I think bytes are just a little harder to use in practice. For example, there's the JPA repository function to get an entity by ID, so if it's bytes the user would need to do some conversion instead of just provide a UUID.

I feel the way this section is written, we're kind of suggesting to users to default to using String UUID but we also tell them how to use Byte UUID if they want. I don't think we're saying that using bytes is problematic.